// ngx-csv-parser/bundles/ngx-csv-parser.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ngx-csv-parser', ['exports', '@angular/core', 'rxjs'], factory) :
    (global = global || self, factory(global['ngx-csv-parser'] = {}, global.ng.core, global.rxjs));
}(this, (function (exports, core, rxjs) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var NgxCSVParserError = /** @class */ (function () {
        function NgxCSVParserError() {
        }
        return NgxCSVParserError;
    }());

    var NgxCsvParser = /** @class */ (function () {
        function NgxCsvParser() {
            this.defaultCSVParserConfig = {
                header: true,
                delimiter: ','
            };
            console.log("\n      *****************************************************\n      ***      For Custom Software/Web development      ***\n      *** Reach out to us at contact@developershive.com ***\n      ***  Support us https://patreon.com/tofiqquadri/  ***\n      ***     Visit us at https://developershive.com    ***\n      *****************************************************\n    ");
        }
        NgxCsvParser.prototype.parse = function (csvFile, config) {
            var _this = this;
            config = __assign(__assign({}, this.defaultCSVParserConfig), config);
            var ngxCSVParserObserver = new rxjs.Observable(function (observer) {
                try {
                    var csvRecords_1 = null;
                    if (_this.isCSVFile(csvFile)) {
                        var reader_1 = new FileReader();
                        reader_1.readAsText(csvFile);
                        reader_1.onload = function () {
                            var csvData = reader_1.result;
                            var csvRecordsArray = _this.csvStringToArray(csvData.trim(), config.delimiter);
                            var headersRow = _this.getHeaderArray(csvRecordsArray);
                            csvRecords_1 = _this.getDataRecordsArrayFromCSVFile(csvRecordsArray, headersRow.length, config);
                            observer.next(csvRecords_1);
                            observer.complete();
                        };
                        reader_1.onerror = function () {
                            _this.badCSVDataFormatErrorHandler(observer);
                        };
                    }
                    else {
                        _this.notCSVFileErrorHandler(observer);
                    }
                }
                catch (error) {
                    _this.unknownCSVParserErrorHandler(observer);
                }
            });
            return ngxCSVParserObserver;
        };
        NgxCsvParser.prototype.csvStringToArray = function (csvDataString, delimiter) {
            var regexPattern = new RegExp(("(\\" + delimiter + "|\\r?\\n|\\r|^)(?:\"((?:\\\\.|\"\"|[^\\\\\"])*)\"|([^\\" + delimiter + "\"\\r\\n]*))"), "gi");
            var matchedPatternArray = regexPattern.exec(csvDataString);
            var resultCSV = [[]];
            while (matchedPatternArray) {
                if (matchedPatternArray[1].length && matchedPatternArray[1] !== delimiter) {
                    resultCSV.push([]);
                }
                var cleanValue = matchedPatternArray[2] ?
                    matchedPatternArray[2].replace(new RegExp("[\\\\\"](.)", "g"), '$1') : matchedPatternArray[3];
                resultCSV[resultCSV.length - 1].push(cleanValue);
                matchedPatternArray = regexPattern.exec(csvDataString);
            }
            return resultCSV;
        };
        NgxCsvParser.prototype.getDataRecordsArrayFromCSVFile = function (csvRecordsArray, headerLength, config) {
            var dataArr = [];
            var headersArray = csvRecordsArray[0];
            var startingRowToParseData = config.header ? 1 : 0;
            for (var i = startingRowToParseData; i < csvRecordsArray.length; i++) {
                var data = csvRecordsArray[i];
                if (data.length === headerLength && config.header) {
                    var csvRecord = {};
                    for (var j = 0; j < data.length; j++) {
                        if ((data[j] === undefined) || (data[j] === null)) {
                            csvRecord[headersArray[j]] = "";
                        }
                        else {
                            csvRecord[headersArray[j]] = data[j].trim();
                        }
                    }
                    dataArr.push(csvRecord);
                }
                else {
                    dataArr.push(data);
                }
            }
            return dataArr;
        };
        NgxCsvParser.prototype.isCSVFile = function (file) {
            return file.name.endsWith('.csv');
        };
        NgxCsvParser.prototype.getHeaderArray = function (csvRecordsArr) {
            var e_1, _a;
            var headers = csvRecordsArr[0];
            var headerArray = [];
            try {
                for (var headers_1 = __values(headers), headers_1_1 = headers_1.next(); !headers_1_1.done; headers_1_1 = headers_1.next()) {
                    var header = headers_1_1.value;
                    headerArray.push(header);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (headers_1_1 && !headers_1_1.done && (_a = headers_1.return)) _a.call(headers_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return headerArray;
        };
        NgxCsvParser.prototype.notCSVFileErrorHandler = function (observer) {
            var ngcCSVParserError = this.errorBuilder('NOT_A_CSV_FILE', 'Selected file is not a csv File Type.', 2);
            observer.error(ngcCSVParserError);
        };
        NgxCsvParser.prototype.unknownCSVParserErrorHandler = function (observer) {
            var ngcCSVParserError = this.errorBuilder('UNKNOWN_ERROR', 'Unknown error. Please refer to official documentation for library usage.', 404);
            observer.error(ngcCSVParserError);
        };
        NgxCsvParser.prototype.badCSVDataFormatErrorHandler = function (observer) {
            var ngcCSVParserError = this.errorBuilder('BAD_CSV_DATA_FORMAT', 'Unable to parse CSV File.', 1);
            observer.error(ngcCSVParserError);
        };
        NgxCsvParser.prototype.errorBuilder = function (type, message, code) {
            var ngcCSVParserError = new NgxCSVParserError();
            ngcCSVParserError.type = type;
            ngcCSVParserError.message = message;
            ngcCSVParserError.code = code;
            return ngcCSVParserError;
        };
        NgxCsvParser.ɵprov = core.ɵɵdefineInjectable({ factory: function NgxCsvParser_Factory() { return new NgxCsvParser(); }, token: NgxCsvParser, providedIn: "root" });
        NgxCsvParser = __decorate([
            core.Injectable({
                providedIn: 'root'
            })
        ], NgxCsvParser);
        return NgxCsvParser;
    }());
    var CSVParserConfig = /** @class */ (function () {
        function CSVParserConfig() {
        }
        return CSVParserConfig;
    }());

    var NgxCsvParserComponent = /** @class */ (function () {
        function NgxCsvParserComponent() {
        }
        NgxCsvParserComponent.prototype.ngOnInit = function () {
        };
        NgxCsvParserComponent = __decorate([
            core.Component({
                selector: 'lib-ngx-csv-parser',
                template: "\n    <p>\n      ngx-csv-parser works!\n    </p>\n  "
            })
        ], NgxCsvParserComponent);
        return NgxCsvParserComponent;
    }());

    var NgxCsvParserModule = /** @class */ (function () {
        function NgxCsvParserModule() {
        }
        NgxCsvParserModule = __decorate([
            core.NgModule({
                declarations: [NgxCsvParserComponent],
                imports: [],
                providers: [NgxCsvParser],
                exports: [NgxCsvParserComponent]
            })
        ], NgxCsvParserModule);
        return NgxCsvParserModule;
    }());

    exports.NgxCSVParserError = NgxCSVParserError;
    exports.NgxCsvParser = NgxCsvParser;
    exports.NgxCsvParserComponent = NgxCsvParserComponent;
    exports.NgxCsvParserModule = NgxCsvParserModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
